File: src/core/geometry/traceBitmap.ts
/**
 * Bitmap → outlines. The shared core under Auto-trace and Create Shapes From
 * Text: an alpha (or luma) plane in, a list of closed contours out, each
 * simplified to the fewest vertices that stay within a pixel tolerance.
 *
 * ## Algorithm
 *
 * 1. Threshold to a binary mask.
 * 2. Border-follow every connected region (Suzuki–Abe style: outer borders
 *    AND hole borders), walking pixel EDGES rather than pixel centres so a
 *    one-pixel-wide stroke still yields a ring with area.
 * 3. Ramer–Douglas–Peucker simplification, tolerance in pixels.
 *
 * Holes come back as separate contours with `hole: true` — a letter's
 * counter, a donut's middle. A consumer that cannot represent holes (a mask
 * path) gets the outer rings; one that can (a Geometry with subpaths) gets
 * both, and the renderer's even-odd fill does the rest.
 *
 * Pure, allocation-light, no DOM. Corner vertices are emitted as corners —
 * smoothing a traced glyph into curves is a separate, opt-in step
 * (`smoothContour`), because a traced rectangle that comes back with rounded
 * corners has lost information the tracer actually had.
 */

export interface TracePoint { x: number; y: number }

export interface TracedContour {
  points: TracePoint[];
  /** True for an inner border (the region's hole), false for an outer one. */
  hole: boolean;
}

export interface TraceOptions {
  /** 0..255 alpha at or above which a pixel is "inside". Default 128. */
  threshold?: number;
  /** RDP tolerance in pixels. Default 1. 0 keeps every edge vertex. */
  tolerance?: number;
  /** Drop contours with fewer pixels of area than this. Default 4. */
  minArea?: number;
}

// ── Binary mask with a one-pixel border so neighbour lookups never bounds-check ──

function toMask(src: Uint8Array | Uint8ClampedArray, w: number, h: number, stride: number, threshold: number): Uint8Array {
  const mw = w + 2;
  const mask = new Uint8Array(mw * (h + 2));
  for (let y = 0; y < h; y++) {
    for (let x = 0; x < w; x++) {
      if (src[(y * w + x) * stride + (stride - 1)]! >= threshold) mask[(y + 1) * mw + (x + 1)] = 1;
    }
  }
  return mask;
}

/**
 * Trace the boundary of the region containing (sx, sy), walking the pixel
 * edges clockwise (outer) or counter-clockwise (hole). Returns vertices on
 * the pixel grid (integer corners) in IMAGE coordinates, and marks every
 * visited boundary pixel in `visited` so the scan does not re-trace it.
 *
 * Edge-following on a 4-connected crack boundary: at each step we stand on a
 * grid corner with a direction; the two pixels ahead decide turn left,
 * straight, or turn right. Classic "square tracing" with the inside kept on
 * the right hand.
 */
const DX = [1, 0, -1, 0];
const DY = [0, 1, 0, -1];

function followEdges(
  mask: Uint8Array,
  mw: number,
  startX: number,
  startY: number,
  startDir: number,
  visited: Uint8Array,
): TracePoint[] {
  // Directions: 0 = +x, 1 = +y, 2 = −x, 3 = −y. Position is a grid CORNER
  // (cx, cy); pixel (px, py) occupies corners [px, px+1] × [py, py+1].
  const inside = (px: number, py: number): boolean => mask[py * mw + px] === 1;
  const pts: TracePoint[] = [];
  let cx = startX, cy = startY, dir = startDir;
  let guard = mw * mw * 4;
  // The start corner is a vertex by construction (the scan picked it where
  // the boundary turns), so it is pushed unconditionally; every later corner
  // is pushed only when the heading changes there — a straight run of edges
  // is one segment, not one vertex per pixel.
  let first = true;
  do {
    // Inside on the RIGHT as we walk. The two pixels ahead of this corner —
    // one on each side of the heading — decide the turn; see AHEAD.
    const ahead = AHEAD[dir]!;
    const R = inside(cx + ahead.rx, cy + ahead.ry);
    const L = inside(cx + ahead.lx, cy + ahead.ly);
    const before = dir;
    if (R && !L) {
      // straight on
    } else if (R && L) {
      dir = (dir + 3) & 3; // concave corner: turn left
    } else {
      // !R: convex corner (or a diagonal touch, which we also treat as a
      // convex turn so two regions meeting at a point stay two regions).
      dir = (dir + 1) & 3;
    }
    if (first || dir !== before) pts.push({ x: cx, y: cy });
    first = false;
    // Mark the inside pixel on our right as a visited boundary pixel.
    const a = AHEAD[dir]!;
    visited[(cy + a.ry) * mw + (cx + a.rx)] = 1;
    cx += DX[dir]!;
    cy += DY[dir]!;
  } while ((cx !== startX || cy !== startY) && --guard > 0);
  return pts;
}

/**
 * For a walker at corner (cx, cy) heading `dir`, the pixel just ahead on its
 * RIGHT and on its LEFT, as corner-relative offsets. Pixel (px,py) has its
 * top-left corner at (px,py).
 */
const AHEAD: ReadonlyArray<{ rx: number; ry: number; lx: number; ly: number }> = [
  { rx: 0, ry: 0, lx: 0, ly: -1 },   // +x: right = below the edge, left = above
  { rx: -1, ry: 0, lx: 0, ly: 0 },   // +y: right = left of the edge
  { rx: -1, ry: -1, lx: -1, ly: 0 }, // −x
  { rx: 0, ry: -1, lx: -1, ly: -1 }, // −y
];

/** Signed area (shoelace); positive = clockwise in y-down image space. */
function signedArea(pts: ReadonlyArray<TracePoint>): number {
  let a = 0;
  for (let i = 0, n = pts.length; i < n; i++) {
    const p = pts[i]!, q = pts[(i + 1) % n]!;
    a += p.x * q.y - q.x * p.y;
  }
  return a / 2;
}

// ── Ramer–Douglas–Peucker on a closed ring ──────────────────────────

function perpDist(p: TracePoint, a: TracePoint, b: TracePoint): number {
  const dx = b.x - a.x, dy = b.y - a.y;
  const len2 = dx * dx + dy * dy;
  if (len2 === 0) return Math.hypot(p.x - a.x, p.y - a.y);
  const t = Math.max(0, Math.min(1, ((p.x - a.x) * dx + (p.y - a.y) * dy) / len2));
  return Math.hypot(p.x - (a.x + t * dx), p.y - (a.y + t * dy));
}

function rdpOpen(pts: ReadonlyArray<TracePoint>, eps: number, out: TracePoint[]): void {
  if (pts.length < 3) { for (const p of pts) out.push(p); return; }
  let maxD = 0, idx = 0;
  const a = pts[0]!, b = pts[pts.length - 1]!;
  for (let i = 1; i < pts.length - 1; i++) {
    const d = perpDist(pts[i]!, a, b);
    if (d > maxD) { maxD = d; idx = i; }
  }
  if (maxD > eps) {
    rdpOpen(pts.slice(0, idx + 1), eps, out);
    out.pop();
    rdpOpen(pts.slice(idx), eps, out);
  } else {
    out.push(a, b);
  }
}

/** Simplify a closed ring: split at the two most distant vertices, RDP each half. */
export function simplifyRing(pts: ReadonlyArray<TracePoint>, eps: number): TracePoint[] {
  if (eps <= 0 || pts.length <= 4) return [...pts];
  // Anchor at the vertex farthest from the first; that pair is a sound split.
  let far = 0, farD = -1;
  for (let i = 1; i < pts.length; i++) {
    const d = Math.hypot(pts[i]!.x - pts[0]!.x, pts[i]!.y - pts[0]!.y);
    if (d > farD) { farD = d; far = i; }
  }
  const half1 = pts.slice(0, far + 1);
  const half2 = [...pts.slice(far), pts[0]!];
  const out: TracePoint[] = [];
  rdpOpen(half1, eps, out);
  out.pop();
  rdpOpen(half2, eps, out);
  out.pop(); // the repeated start
  // Drop collinear survivors the split may have left at the seams.
  return out.filter((p, i, arr) => {
    const a = arr[(i + arr.length - 1) % arr.length]!, b = arr[(i + 1) % arr.length]!;
    return perpDist(p, a, b) > 1e-6;
  });
}

/**
 * Trace every region in an 8-bit plane.
 *
 * `stride` is the bytes per pixel and the value is read from the LAST byte of
 * each pixel — 1 for a luma/alpha plane, 4 for RGBA (alpha). Coordinates are
 * image pixels with (0,0) at the top-left corner of the top-left pixel.
 */
export function traceBitmap(
  src: Uint8Array | Uint8ClampedArray,
  w: number,
  h: number,
  stride: 1 | 4 = 1,
  opts: TraceOptions = {},
): TracedContour[] {
  const threshold = opts.threshold ?? 128;
  const tolerance = opts.tolerance ?? 1;
  const minArea = opts.minArea ?? 4;
  const mw = w + 2;
  const mask = toMask(src, w, h, stride, threshold);
  const visitedOuter = new Uint8Array(mask.length);
  const visitedHole = new Uint8Array(mask.length);
  const out: TracedContour[] = [];

  // Raster scan. An OUTER border starts where a row goes outside→inside with
  // the pixel above the start corner outside; a HOLE border where it goes
  // inside→outside. Each boundary pixel is visited once per border kind.
  for (let py = 1; py <= h; py++) {
    for (let px = 1; px <= w; px++) {
      const here = mask[py * mw + px] === 1;
      const left = mask[py * mw + px - 1] === 1;
      // `!visitedHole` too: the inside pixel just right of a hole also reads
      // as "inside with outside on its left", but it is a hole-border pixel
      // and the hole walk — which started at the hole's left edge earlier on
      // this same row — has already marked it.
      if (here && !left && !visitedOuter[py * mw + px] && !visitedHole[py * mw + px]) {
        // Outer border: from the pixel's top-left corner heading +x along its
        // top edge — the pixel itself is then on the walker's right.
        const pts = followEdges(mask, mw, px, py, 0, visitedOuter);
        visitedOuter[py * mw + px] = 1;
        const ring = pts.map((p) => ({ x: p.x - 1, y: p.y - 1 }));
        if (Math.abs(signedArea(ring)) < minArea) continue;
        out.push({ points: simplifyRing(ring, tolerance), hole: false });
      } else if (!here && left && !visitedHole[py * mw + px - 1]) {
        // Hole border: the pixel to the left is inside, this one is a hole
        // (or the outside). Only a hole if the run is enclosed: walking the
        // border with inside on the right and checking orientation tells us.
        // Heading +y down the edge between them keeps the inside pixel (to
        // the west) on the right.
        const pts = followEdges(mask, mw, px, py, 1, visitedHole);
        visitedHole[py * mw + px - 1] = 1;
        const ring = pts.map((p) => ({ x: p.x - 1, y: p.y - 1 }));
        const area = signedArea(ring);
        if (Math.abs(area) < minArea) continue;
        // The outside of the whole image traced from this seed comes back as
        // a counter-clockwise ring around the region's OUTER edge — the same
        // ring the outer pass emits. Skip those by orientation: a hole walked
        // with inside-on-right is clockwise-negative in y-down space.
        if (area > 0) continue;
        out.push({ points: simplifyRing(ring, tolerance), hole: true });
      }
    }
  }
  return out;
}

/**
 * Turn a polygon into a smooth closed Bézier (Catmull–Rom tangents), for the
 * consumers that want a traced letterform to read as a curve rather than a
 * polyline. `tension` 0 = straight segments, 1 = fully rounded.
 */
export function smoothContour(
  pts: ReadonlyArray<TracePoint>,
  tension = 0.5,
): Array<{ x: number; y: number; inX: number; inY: number; outX: number; outY: number }> {
  const n = pts.length;
  const k = tension / 3; // Catmull-Rom → Bézier handle factor at tension 1 is 1/6 per side; /3 reads better for traced art
  return pts.map((p, i) => {
    const prev = pts[(i + n - 1) % n]!, next = pts[(i + 1) % n]!;
    const tx = (next.x - prev.x) * k, ty = (next.y - prev.y) * k;
    return { x: p.x, y: p.y, inX: p.x - tx, inY: p.y - ty, outX: p.x + tx, outY: p.y + ty };
  });
}

File: native/engine/src/scene/trace_bitmap.cpp
// src/core/geometry/traceBitmap.ts, call for call (see extrude_mesh.hpp).
#include <algorithm>
#include <array>
#include <cmath>
#include <cstdint>
#include <numbers>
#include <span>
#include <vector>

#include "extrude_mesh.hpp"
#include "jsmath.hpp"

namespace premation::scene::mesh {
namespace {

namespace jm = motion::js;

double hyp2(double a, double b) noexcept {
  const std::array<double, 2> v{a, b};
  return jm::hypot(std::span<const double>(v));
}

struct Ahead {
  int rx;
  int ry;
  int lx;
  int ly;
};
constexpr std::array<int, 4> kDX{1, 0, -1, 0};
constexpr std::array<int, 4> kDY{0, 1, 0, -1};
constexpr std::array<Ahead, 4> kAhead{{
    {0, 0, 0, -1},
    {-1, 0, 0, 0},
    {-1, -1, -1, 0},
    {0, -1, -1, -1},
}};

std::vector<std::uint8_t> to_mask(std::span<const std::uint8_t> src, int w, int h, int stride, double threshold) {
  const int mw = w + 2;
  std::vector<std::uint8_t> mask(static_cast<std::size_t>(mw) * static_cast<std::size_t>(h + 2));
  for (int y = 0; y < h; ++y) {
    for (int x = 0; x < w; ++x) {
      const auto at = (static_cast<std::size_t>((y * w) + x) * static_cast<std::size_t>(stride)) + static_cast<std::size_t>(stride - 1);
      if (static_cast<double>(src[at]) >= threshold) mask[static_cast<std::size_t>(((y + 1) * mw) + (x + 1))] = 1;
    }
  }
  return mask;
}

std::vector<Pt2> follow_edges(const std::vector<std::uint8_t>& mask, int mw, int startX, int startY, int startDir,
                              std::vector<std::uint8_t>& visited) {
  const auto inside = [&](int px, int py) { return mask[static_cast<std::size_t>((py * mw) + px)] == 1; };
  std::vector<Pt2> pts;
  int cx = startX;
  int cy = startY;
  int dir = startDir;
  long long guard = static_cast<long long>(mw) * mw * 4;
  bool first = true;
  do {
    const Ahead& ahead = kAhead[static_cast<std::size_t>(dir)];
    const bool R = inside(cx + ahead.rx, cy + ahead.ry);
    const bool L = inside(cx + ahead.lx, cy + ahead.ly);
    const int before = dir;
    if (R && !L) {
      // straight on
    } else if (R && L) {
      dir = (dir + 3) & 3;
    } else {
      dir = (dir + 1) & 3;
    }
    if (first || dir != before) pts.push_back({static_cast<double>(cx), static_cast<double>(cy)});
    first = false;
    const Ahead& a = kAhead[static_cast<std::size_t>(dir)];
    visited[static_cast<std::size_t>(((cy + a.ry) * mw) + (cx + a.rx))] = 1;
    cx += kDX[static_cast<std::size_t>(dir)];
    cy += kDY[static_cast<std::size_t>(dir)];
  } while ((cx != startX || cy != startY) && --guard > 0);
  return pts;
}

double trace_area(std::span<const Pt2> pts) noexcept {
  double a = 0;
  const std::size_t n = pts.size();
  for (std::size_t i = 0; i < n; ++i) {
    const Pt2 p = pts[i];
    const Pt2 q = pts[(i + 1) % n];
    a += (p.x * q.y) - (q.x * p.y);
  }
  return a / 2;
}

double perp_dist(Pt2 p, Pt2 a, Pt2 b) noexcept {
  const double dx = b.x - a.x;
  const double dy = b.y - a.y;
  const double len2 = (dx * dx) + (dy * dy);
  if (len2 == 0) return hyp2(p.x - a.x, p.y - a.y);
  const std::array<double, 2> inner{0.0, (((p.x - a.x) * dx) + ((p.y - a.y) * dy)) / len2};
  const double lo = jm::max_of(std::span<const double>(inner));
  const std::array<double, 2> outer{1.0, lo};
  const double t = jm::min_of(std::span<const double>(outer));
  return hyp2(p.x - (a.x + (t * dx)), p.y - (a.y + (t * dy)));
}

void rdp_open(std::span<const Pt2> pts, double eps, std::vector<Pt2>& out) {
  if (pts.size() < 3) {
    for (const Pt2& p : pts) out.push_back(p);
    return;
  }
  double maxD = 0;
  std::size_t idx = 0;
  const Pt2 a = pts.front();
  const Pt2 b = pts.back();
  for (std::size_t i = 1; i + 1 < pts.size(); ++i) {
    const double d = perp_dist(pts[i], a, b);
    if (d > maxD) {
      maxD = d;
      idx = i;
    }
  }
  if (maxD > eps) {
    rdp_open(pts.subspan(0, idx + 1), eps, out);
    out.pop_back();
    rdp_open(pts.subspan(idx), eps, out);
  } else {
    out.push_back(a);
    out.push_back(b);
  }
}

}  // namespace

std::vector<Pt2> simplify_ring(std::span<const Pt2> pts, double eps) {
  if (eps <= 0 || pts.size() <= 4) return {pts.begin(), pts.end()};
  std::size_t far = 0;
  double farD = -1;
  for (std::size_t i = 1; i < pts.size(); ++i) {
    const double d = hyp2(pts[i].x - pts[0].x, pts[i].y - pts[0].y);
    if (d > farD) {
      farD = d;
      far = i;
    }
  }
  const std::vector<Pt2> half1(pts.begin(), pts.begin() + static_cast<std::ptrdiff_t>(far + 1));
  std::vector<Pt2> half2(pts.begin() + static_cast<std::ptrdiff_t>(far), pts.end());
  half2.push_back(pts[0]);
  std::vector<Pt2> out;
  rdp_open(half1, eps, out);
  out.pop_back();
  rdp_open(half2, eps, out);
  out.pop_back();
  std::vector<Pt2> kept;
  const std::size_t n = out.size();
  for (std::size_t i = 0; i < n; ++i) {
    const Pt2 a = out[(i + n - 1) % n];
    const Pt2 b = out[(i + 1) % n];
    if (perp_dist(out[i], a, b) > 1e-6) kept.push_back(out[i]);
  }
  return kept;
}

std::vector<TracedContour> trace_bitmap(std::span<const std::uint8_t> src, int w, int h, int stride, const TraceOptions& opts) {
  const int mw = w + 2;
  const std::vector<std::uint8_t> mask = to_mask(src, w, h, stride, opts.threshold);
  std::vector<std::uint8_t> visitedOuter(mask.size());
  std::vector<std::uint8_t> visitedHole(mask.size());
  std::vector<TracedContour> out;
  const auto at = [&](int x, int y) { return static_cast<std::size_t>((y * mw) + x); };
  for (int py = 1; py <= h; ++py) {
    for (int px = 1; px <= w; ++px) {
      const bool here = mask[at(px, py)] == 1;
      const bool left = mask[at(px - 1, py)] == 1;
      if (here && !left && visitedOuter[at(px, py)] == 0 && visitedHole[at(px, py)] == 0) {
        std::vector<Pt2> ring = follow_edges(mask, mw, px, py, 0, visitedOuter);
        visitedOuter[at(px, py)] = 1;
        for (Pt2& p : ring) p = {p.x - 1, p.y - 1};
        if (std::abs(trace_area(ring)) < opts.minArea) continue;
        out.push_back({simplify_ring(ring, opts.tolerance), false});
      } else if (!here && left && visitedHole[at(px - 1, py)] == 0) {
        std::vector<Pt2> ring = follow_edges(mask, mw, px, py, 1, visitedHole);
        visitedHole[at(px - 1, py)] = 1;
        for (Pt2& p : ring) p = {p.x - 1, p.y - 1};
        const double area = trace_area(ring);
        if (std::abs(area) < opts.minArea) continue;
        if (area > 0) continue;
        out.push_back({simplify_ring(ring, opts.tolerance), true});
      }
    }
  }
  return out;
}

std::vector<BezPt> smooth_contour(std::span<const Pt2> pts, double tension, std::optional<double> cornerAngleDeg) {
  const std::size_t n = pts.size();
  const double k = tension / 3;
  std::vector<BezPt> out;
  out.reserve(n);
  if (!cornerAngleDeg) {
    for (std::size_t i = 0; i < n; ++i) {
      const Pt2 p = pts[i];
      const Pt2 prev = pts[(i + n - 1) % n];
      const Pt2 next = pts[(i + 1) % n];
      const double tx = (next.x - prev.x) * k;
      const double ty = (next.y - prev.y) * k;
      out.push_back({p.x, p.y, p.x - tx, p.y - ty, p.x + tx, p.y + ty});
    }
    return out;
  }
  const double cosCorner = jm::cos(*cornerAngleDeg * std::numbers::pi / 180);
  const std::array<double, 2> fv{1.0 / 3, k * 2};
  const double f = jm::min_of(std::span<const double>(fv));
  for (std::size_t i = 0; i < n; ++i) {
    const Pt2 p = pts[i];
    const Pt2 prev = pts[(i + n - 1) % n];
    const Pt2 next = pts[(i + 1) % n];
    const double ax = p.x - prev.x;
    const double ay = p.y - prev.y;
    const double bx = next.x - p.x;
    const double by = next.y - p.y;
    const double la = hyp2(ax, ay);
    const double lb = hyp2(bx, by);
    const BezPt sharp{p.x, p.y, p.x, p.y, p.x, p.y};
    if (la < 1e-6 || lb < 1e-6) {
      out.push_back(sharp);
      continue;
    }
    if (((ax * bx) + (ay * by)) / (la * lb) <= cosCorner) {
      out.push_back(sharp);
      continue;
    }
    const double cx = next.x - prev.x;
    const double cy = next.y - prev.y;
    double lc = hyp2(cx, cy);
    if (lc == 0 || std::isnan(lc)) lc = 1;
    const double ux = cx / lc;
    const double uy = cy / lc;
    out.push_back({p.x, p.y, p.x - (ux * la * f), p.y - (uy * la * f), p.x + (ux * lb * f), p.y + (uy * lb * f)});
  }
  return out;
}

}  // namespace premation::scene::mesh
